README.md
# Ubuntu Web VNC Container 

Run Ubuntu Web VNC within Docker with one command!

**Credit**: based on [ConSol](https://github.com/ConSol/docker-headless-vnc-container) 's Version (achieved)

**Current TODOs**: [changelog, TODOs, and Help Wanted](changelog.md)

**Features**:

* Desktop environment: [**Xfce4**](http://www.xfce.org)
* VNCServer: latest [**TigerVNC**](https://github.com/TigerVNC/tigervnc)
  * which doesn't rely on unreliable `xstartup` to start desktop session!
* HTML5 VNC client: [**noVNC**](https://github.com/novnc/noVNC)
* Software:
  * Mozilla Firefox
  * Chromium (not work for `ARM`, a known problem of Chrome?)
  * Sublime
  * Docker
  * Chinese Input (`Ctrl` + `Space`)
 
## OS & UI sessions:

* `kaiyhou/ubuntu-web-vnc`: 
  * __Ubuntu 18.04__ with `Xfce4` UI session
      * X86-64
      * ARM-64
      
## Usage

- Run command with mapping to local port `8002` (vnc web access):

      bash run-container.sh [password]
      # visit: http://Your_IP|Domain:8002
      

- Advanced usage
    ```bash
    docker run -d \
      --privileged \
      --restart=on-failure:10 \
      --shm-size=512m \
      -v /var/run/docker.sock:/var/run/docker.sock \
      -v /etc/timezone:/etc/timezone:ro \
      -p 8001:5901 -p 8002:6901 \
      -e VNC_PW=VNC_PASSWORD \
      -e VNC_RESOLUTION=1600x900 \
      -e VNC_CLIENT=false \
      -v ~/vnc-data:/headless/share \
      --name ubuntu-web-vnc \
      kaiyhou/ubuntu-web-vnc
    ```

- Build an image from scratch:

      # for single arch
      bash build-dev.sh
      
      # for multi arch (amd64, arm64) with proper tags
      bash build-multi-arch.sh

# Connect & Control
If the container is started like mentioned above, connect via one of these options:

* __noVNC HTML5 full client__: [`http://localhost:8002/vnc.html`](http://localhost:8002/vnc.html), default password: `VNC_PASSWORD`
  * With clipboard and Scale at side panel!
* __noVNC HTML5 lite client__: [`http://localhost:8002/?password=VNC_PASSWORD`](http://localhost:8002/?password=VNC_PASSWORD) 

* __VNC viewer Client__: `localhost:8001`, default password: `VNC_PASSWORD`, disabled by defalut
  * Requirement: `-e VNC_CLIENT=true` in `docker run` command

## Hints

### 0) Security Consideration: best practice
  In short, disable vnc client option (disabled by default) and use `Caddy` server to proxy the web VNC by `https`.
  * Disable VNC Client: `-e VNC_CLIENT=false \`
  * Set VNC Port to localhost only: `-p 127.0.0.1:8001:5901 -p 127.0.0.1:8002:6901 \`
  * Use [`Caddy`](https://caddyserver.com/docs/) Server to reverse proxy Web VNC with https
    * ```
      # /etc/caddy/Caddyfile
      vnc.your_domain.com {
        reverse_proxy localhost:8002
      }
      ```
  * Access Web VNC via `https`: `https://vnc.your_domain.com`

### 1) Extend a Image with your own software
All images run as non-root user per default, so if you want to extend the image and install software, you have to switch back to the `root` user:

```bash
## Custom Dockerfile
FROM kaiyhou/ubuntu-web-vnc

# Switch to root user to install additional software
USER 0

## Install a gedit
RUN apt-get install -y gedit \
    && apt-get clean all

## switch back to default user
USER ubuntu
```

### 2) Using root (user id `0`)
Add the `--user` flag to your docker run command:

    docker run -it --user 0 ... kaiyhou/ubuntu-web-vnc

### 2) Using user and group id of host system
Add the `--user` flag to your docker run command:

    docker run -it --user $(id -u):$(id -g) ... kaiyhou/ubuntu-web-vnc

### 3) Override VNC environment variables
The following VNC environment variables can be overwritten at the `docker run` phase to customize your desktop environment inside the container:
* `VNC_COL_DEPTH`, default: `24`
* `VNC_RESOLUTION`, default: `1600x900`
* `VNC_PW`, default: `VNC_PASSWORD`

#### 3.1) Example: Override the VNC password
Simply overwrite the value of the environment variable `VNC_PW`. For example in
the docker run command:

    docker run -it -e VNC_PW=my-pw ... kaiyhou/ubuntu-web-vnc

#### 3.2) Example: Override the VNC resolution
Simply overwrite the value of the environment variable `VNC_RESOLUTION`. For example in
the docker run command:

    docker run -it -e VNC_RESOLUTION=800x600 ... kaiyhou/ubuntu-web-vnc
    
### 4) View only VNC
Since version `1.2.0` it's possible to prevent unwanted control via VNC. Therefore you can set the environment variable `VNC_VIEW_ONLY=true`. If set, the startup script will create a random password for the control connection and use the value of `VNC_PW` for view only connection over the VNC connection.

     docker run -it -e VNC_VIEW_ONLY=true ... kaiyhou/ubuntu-web-vnc

### 5) Chromium crashes with high VNC_RESOLUTION 
The problem there is the too small `/dev/shm` size in the container. Except the solution below, we can also use `-v /dev/shm:/dev/shm`

    docker run --shm-size=256m -it VNC_RESOLUTION=1920x1080 ... kaiyhou/ubuntu-web-vnc

### 6) Chinese Input Usage

 * `Ctrl` + `Space` switch to Chinese input: `fcitx-rime`.
 * In the `rime` input, simple click left `shift` to switch between `Chinese` and `English`
 * Traditional/Simplification Chinese, and other features switch: `Ctrl` + `~`
 * Issues: 
   * `Fcitx Configuration` application in xfce4 application menu cannot link to `fcitx` server automatically.
   * Solution: `killall fcitx` then `fcitx`. Then, you can config fcitx via GUI.

  

## Contributors

At this point we want to thank all contributors, which helped to move this great project by submitting code, writing documentation, or adapting other tools to play well together with the docker headless container.

* [Tobias Schneck](https://github.com/toschneck)
* [Robert Bohne](https://github.com/rbo) - IceWM images
* [hsiaoyi0504](https://github.com/hsiaoyi0504) - PR [#66](https://github.com/ConSol/docker-headless-vnc-container/pull/66)
* [dmhumph](https://github.com/dmhumph) - PR [#44](https://github.com/ConSol/docker-headless-vnc-container/issue/44) 
* [Simon Hofmann](https://github.com/s1hofmann)

## Changelog

The current changelog is provided here: **[changelog.md](./changelog.md)**



changelog.md
## Changelog after Kaiyu took care of it

TODOs
 * Update to Ubuntu 20.04
 * Add domain name to the noVNC title
 * Git Action for auto build and push
 * Sound (we can here sound in vnc.html when delete a file)
    * tighervnc does not support sound yet: https://github.com/TigerVNC/tigervnc/issues/396
    * bypass: pulseaudio with SSH forward
    * one solution https://github.com/wu191287278/noVNC-audio
    * xrdp supports sounds
 * Do we really need the `libnss_wrapper`?
 * Update noVNC client to the latest version
 * Install Qogir Icon theme via Network/apt
   * Or using Zip to install Qogir and Papirus, instead of APT and COPY

Achieved
* Chrome not work on ARM64 Oracle Linux (seems to be a problem of chromium)

2021-10-25
 * Re-organize Dockerfile for easier developer
 * Remove supervisor
 * Chinese Input: rime by default: `Ctrl` + `Space`
 * A option to enable VNC Client: `-e VNC_CLIENT=false`
 * Tag to version 2.0.0

2021-10-23
 * Rename project to `ubuntu-web-vnc`
    
2021-09-21
 * Figured out that tigerVNC don't rely on `xstartup` file to config the desktop
 * For more information:
    * https://github.com/TigerVNC/tigervnc/blob/master/unix/vncserver/HOWTO.md
    * https://github.com/TigerVNC/tigervnc/issues/1271
    * Desktop path: `/usr/share/xsessions`
 * Figured out the only beep sound is coming from the novnc web page:
    * https://github.com/novnc/noVNC/tree/master/app/sounds

2021-09-10
 * Remove redundant content in Qogir theme
 * Support ARM64 (with buildx script)
 * Update run script with password input support
 * ARM 64 Support Firefox

2021-09-09
 * update to Ubuntu 18.04
 * add user ubuntu with sudo permission
 * install sublime and docker
 * sync the timezone from the host
 * set up xfce4 theme and icons, and many more details of the UI
 * BUG SOLVED: two xfce4 sessions, two xfce4-panels, etc...


## Consol's Changelog of the Docker headless VNC images

### Version 1.4.0:
* added Github issue and pull request templates
* added install script for custom fonts ([#83](https://github.com/ConSol/docker-headless-vnc-container/issues/83)))
* fixed problem that VNC password was not properly set after doing a Docker commit ([#88](https://github.com/ConSol/docker-headless-vnc-container/issues/88))
* fixed some typos ([#87](https://github.com/ConSol/docker-headless-vnc-container/issues/87))
* added 'how-to-release' info ([#75](https://github.com/ConSol/docker-headless-vnc-container/issues/75))
* added Travis build ([#74](https://github.com/ConSol/docker-headless-vnc-container/issues/74))

### Version 1.3.0:
* change default USER to `1000` ([#61](https://github.com/ConSol/docker-headless-vnc-container/issues/61))
* refactor vnc startup script ([#73](https://github.com/ConSol/docker-headless-vnc-container/issues/73))
  * add help option `--help`
  * ensure correct forwarding of shutdown signals
  * add "DEBUG" mode and move all log output to this mode
  * update README.md
* merge pull request from:
  * [hsiaoyi0504](https://github.com/hsiaoyi0504) update noVNC to [v1.0.0](https://github.com/novnc/noVNC/releases/tag/v1.0.0) ([#66](https://github.com/ConSol/docker-headless-vnc-container/pull/66))
* add example for [Kubernetes usage](./kubernetes/README.md) ([#71](https://github.com/ConSol/docker-headless-vnc-container/issues/71)) 
* remove verbose output by default from `set_user_permissions.sh` script
 
### Version 1.2.3:

* start no_vnc with localhost instead of VNC_IP to ensure startup on docker bridge mode ([#41](https://github.com/ConSol/docker-headless-vnc-container/issue/41)) 
* add xfce4-terminal to enable copy and paste ([#55](https://github.com/ConSol/docker-headless-vnc-container/issue/55))
* add save mode for tag_image.sh script 
* disable trigger latest Dockerhub build -> use tag_image.sh script 
* use correct python-numpy packages 
* add "Known Issues" documentation: [Chromium crashes with high VNC_RESOLUTION](https://github.com/ConSol/docker-headless-vnc-container#51-chromium-crashes-with-high-vnc_resolution-53)

### Version 1.2.2
* disable automatic firefox updates ([#45](https://github.com/ConSol/docker-headless-vnc-container/issue/45))
* fix wrong env usage of `$HOME` ([#47](https://github.com/ConSol/docker-headless-vnc-container/issue/47))
* fix hanging vnc handshake if container is offline ([#50](https://github.com/ConSol/docker-headless-vnc-container/issue/50))
* Merge pull request from:
    * [dmhumph](https://github.com/dmhumph): Update OpenShift instructions for OCP 3.5 or greater ([#44](https://github.com/ConSol/docker-headless-vnc-container/issue/44)) 

### Version 1.2.1
* Centos7: use temporally epel-testing repo to fix chrome error `libpng warning: Application built with libpng-1.6.22 but running with 1.5.13` in version `60.0.3112.113-2.el7`
* Fix broken `docker stop/start` ([#40](https://github.com/ConSol/docker-headless-vnc-container/issues/40))

### Version 1.2.0
* Use fixed firefox version `45.9.0esr` ([#39](https://github.com/ConSol/docker-headless-vnc-container/issues/39), [#20](https://github.com/ConSol/docker-headless-vnc-container/issues/20), [#16](https://github.com/ConSol/docker-headless-vnc-container/issues/16)) 
  * fix memory consumption problems in comparison to current firefox 55
  * tested on long term container issues without crashes 
* Add en_US.UTF-8 as default locale ([#27](https://github.com/ConSol/docker-headless-vnc-container/issues/27)) 
* Add enviroment variable `VNC_VIEW_ONLY=true` for view only VNC connections ([#25](https://github.com/ConSol/docker-headless-vnc-container/issues/25))
  * Update to tigervnc version `1.8`
* Use a more stable command for ip determination (PR [#31](https://github.com/ConSol/docker-headless-vnc-container/issues/31), thx to @arnaudveron) 
* Support empty `CMD` values in startup script ([#32](https://github.com/ConSol/docker-headless-vnc-container/issues/32)) 

run-container.sh
#!/bin/bash

PWD="VNC_PASSWORD"
if [ "$1" ] ;then
  PWD="$1"
fi

VNCPT="8001"
if [ "$2" ] ;then
  VNCPT="$2"
fi

WEBPT="8002"
if [ "$3" ] ;then
  WEBPT="$3"
fi

docker run -d \
  --privileged \
  --restart=on-failure:10 \
  --shm-size=512m \
  -v /var/run/docker.sock:/var/run/docker.sock \
  -v /etc/timezone:/etc/timezone:ro \
  -p "$VNCPT":5901 -p "$WEBPT":6901 \
  -e VNC_PW="$PWD" \
  -e VNC_RESOLUTION=1600x900 \
  -e VNC_CLIENT=false \
  -v ~/vnc-data:/headless/share \
  --name ubuntu-web-vnc \
  kaiyhou/ubuntu-web-vnc

## --user 0: use root account, optional
## 5091: VNC, 6091: browser VNC
## timezone: share the timezone from host
## --shm-size=256m: avoid chrome crush
##                  another option is -v /dev/shm:/dev/shm
##                  https://github.com/ConSol/docker-headless-vnc-container/issues/53l